crates/energy/src/graph.rs

```diff
@@ -0,0 +1,208 @@
+use std::sync::Mutex;
+
+use bevy::prelude::*;
+use de_core::gamestate::GameState;
+use de_core::projection::ToFlat;
+use de_core::state::AppState;
+use de_index::SpatialQuery;
+use de_spawner::DespawnedComponentsEvent;
+use parry3d::bounding_volume::Aabb;
+use parry3d::math::Point;
+use tinyvec::TinyVec;
+
+use crate::Battery;
+
+/// The max distance (in meters) between two entities for them to be consider neighbors in the graph
+const MAX_DISTANCE: f32 = 10.0;
+/// Minimum distance squared traveled by an object to update its nearby units.
+const MIN_DISTANCE_FOR_UPDATE: f32 = 0.5;
+
+pub(crate) struct GraphPlugin;
+
+impl Plugin for GraphPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_systems(
+            PostUpdate,
+            spawn_graph_components.run_if(in_state(AppState::InGame)),
+        )
+        .add_systems(
+            Update,
+            remove_old_nodes.run_if(in_state(GameState::Playing)),
+        )
+        .add_systems(
+            FixedUpdate,
+            (update_nearby.in_set(GraphSystemSet::UpdateNearby),)
+                .run_if(in_state(AppState::InGame)),
+        );
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+}
+
+/// wrapped entity to allow for default values (so we can work with TinyVec)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct NearbyEntity(Entity);
+
+impl Default for NearbyEntity {
+    fn default() -> Self {
+        Self(Entity::PLACEHOLDER)
+    }
+}
+
+impl From<NearbyEntity> for Entity {
+    fn from(val: NearbyEntity) -> Self {
+        val.0
+    }
+}
+
+impl From<Entity> for NearbyEntity {
+    fn from(entity: Entity) -> Self {
+        Self(entity)
+    }
+}
+
+/// The energy grid member component is used to store the energy grid member entities.
+#[derive(Component, Debug, Clone)]
+pub struct EnergyGridMember;
+
+/// The nearby units component is used to store the nearby entities of an entity.
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits {
+    units: TinyVec<[NearbyEntity; 16]>,
+    last_pos: Option<Vec2>,
+}
+
+impl NearbyUnits {
+    fn remove_matching(&mut self, entity: NearbyEntity) {
+        let index = match self.units.iter().position(|e| *e == entity) {
+            Some(index) => index,
+            None => return,
+        };
+
+        self.units.swap_remove(index);
+    }
+
+    pub fn len(&self) -> usize {
+        self.units.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.units.is_empty()
+    }
+}
+
+/// This system inserts newly spawned units into the energy grid.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Battery>>,
```

Comment:
That was the original code see: https://github.com/DigitalExtinction/Game/pull/649#discussion_r1314310563. I think it would be a better idea just to fix the system order as its more logical that each battery has a connection to the graph.

Reply:
LOL, I am disputing my own idea. Now it leads to the delay so I would try to solve it one way or another (possibly with an explanatory comment why it is needed).